File: packages/ocel/src/env-source.test.ts
import { describe, expect, it } from "vitest";

import { defineConfig } from "./config.js";
import { exec, infisical } from "./env-source.js";

const roundTrip = (value: unknown) => JSON.parse(JSON.stringify(value));

describe("infisical", () => {
  it("serialises its options keyed by infisical", () => {
    expect(roundTrip(infisical({ project: "p-1", environment: "prod" }))).toEqual({
      infisical: { project: "p-1", environment: "prod" },
    });
  });

  it("carries the machine identity a deployed tier signs in as", () => {
    expect(
      roundTrip(
        infisical({
          project: "p-1",
          environment: "prod",
          path: "/acme",
          write: "missing",
          auth: {
            universal: {
              clientId: { var: "INFISICAL_CLIENT_ID" },
              clientSecret: { var: "INFISICAL_CLIENT_SECRET" },
            },
          },
        }),
      ),
    ).toEqual({
      infisical: {
        project: "p-1",
        environment: "prod",
        path: "/acme",
        write: "missing",
        auth: {
          universal: {
            clientId: { var: "INFISICAL_CLIENT_ID" },
            clientSecret: { var: "INFISICAL_CLIENT_SECRET" },
          },
        },
      },
    });
  });
});

describe("exec", () => {
  it("serialises the command and the format it prints, keyed by exec", () => {
    expect(roundTrip(exec({ command: ["op", "inject", "{folder}"], format: "dotenv" }))).toEqual({
      exec: { command: ["op", "inject", "{folder}"], format: "dotenv" },
    });
  });
});

describe("an envSource built from the helpers", () => {
  it("serialises every tier it names", () => {
    const config = defineConfig({
      slug: "test-app",
      envSource: {
        production: infisical({
          project: "p-1",
          environment: "prod",
          auth: { aws: { identityId: "ident" } },
        }),
        preview: "builtin",
        dev: exec({ command: ["./scripts/dev-env.sh"], format: "json" }),
      },
    });

    expect(roundTrip(config.envSource)).toEqual({
      production: {
        infisical: { project: "p-1", environment: "prod", auth: { aws: { identityId: "ident" } } },
      },
      preview: "builtin",
      dev: { exec: { command: ["./scripts/dev-env.sh"], format: "json" } },
    });
  });
});

File: cli/internal/projectconfig/envsource_test.go
package projectconfig

import (
	"context"
	"path/filepath"
	"slices"
	"testing"

	"github.com/ocelhq/ocel/pkg/providerkit/envsource"
)

func TestResolveLeavesEveryTierOnItsDefaultSource(t *testing.T) {
	dir := t.TempDir()
	write(t, filepath.Join(dir, DefaultFileName), `{"slug":"acme"}`)

	cfg, err := Resolve(context.Background(), dir, "")
	if err != nil {
		t.Fatalf("resolve: %v", err)
	}
	if cfg.EnvSource.Production.Kind != envsource.Builtin || cfg.EnvSource.Preview.Kind != envsource.Builtin {
		t.Fatalf("deployed tiers = %+v, want builtin", cfg.EnvSource)
	}
	if cfg.EnvSource.Dev.Kind != envsource.Dotenv {
		t.Fatalf("dev = %+v, want dotenv", cfg.EnvSource.Dev)
	}
}

func TestResolveFillsAnInfisicalSourceWithItsDefaults(t *testing.T) {
	dir := t.TempDir()
	write(t, filepath.Join(dir, DefaultFileName), `{"slug":"acme","envSource":{
		"production":{"infisical":{"project":"p-1","environment":"prod","auth":{"universal":{"clientId":{"var":"ID"},"clientSecret":{"var":"SECRET"}}}}},
		"preview":{"infisical":{"project":"p-1","environment":"staging","path":"/acme/","host":"https://infisical.example.com/","write":"missing","auth":{"aws":{"identityId":"ident"}}}},
		"dev":{"exec":{"command":["op","run","{folder}"],"format":"json"}}
	}}`)

	cfg, err := Resolve(context.Background(), dir, "")
	if err != nil {
		t.Fatalf("resolve: %v", err)
	}
	production := cfg.EnvSource.Production
	if production.Kind != envsource.Infisical || production.Infisical == nil {
		t.Fatalf("production = %+v", production)
	}
	want := envsource.InfisicalOptions{
		Project:     "p-1",
		Environment: "prod",
		Path:        "/",
		Host:        envsource.InfisicalCloud,
		Write:       envsource.WriteNever,
		Auth:        envsource.InfisicalAuth{Method: envsource.AuthUniversal, ClientIDVar: "ID", ClientSecretVar: "SECRET"},
	}
	if *production.Infisical != want {
		t.Fatalf("production infisical = %+v, want %+v", *production.Infisical, want)
	}
	preview := cfg.EnvSource.Preview.Infisical
	if preview.Path != "/acme" || preview.Host != "https://infisical.example.com" || preview.Write != envsource.WriteMissing ||
		preview.Auth != (envsource.InfisicalAuth{Method: envsource.AuthAWS, IdentityID: "ident"}) {
		t.Fatalf("preview infisical = %+v", preview)
	}
	dev := cfg.EnvSource.Dev
	if dev.Kind != envsource.Exec || dev.Exec == nil || !slices.Equal(dev.Exec.Command, []string{"op", "run", "{folder}"}) || dev.Exec.Format != envsource.FormatJSON {
		t.Fatalf("dev = %+v", dev)
	}
}
